cloudsync/youtube.py

```diff
@@ -0,0 +1,256 @@
+""" YouTube API interface"""
+import http
+import time
+import logging
+from tempfile import NamedTemporaryFile
+
+import boto3
+import httplib2
+import oauth2client
+
+from django.conf import settings
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+from googleapiclient.http import MediaFileUpload
+
+from ui.encodings import EncodingNames
+
+log = logging.getLogger(__name__)
+
+
+def resumable_upload(request, max_retries=10):
+    """
+    Upload a video to YouTube and resume on failure up to 10 times, adapted from YouTube API example.
+    To use resumable media you must use a MediaFileUpload object and flag it as a resumable upload.
+    You then repeatedly call next_chunk() on the googleapiclient.http.HttpRequest object until the
+    upload is complete.
+
+    Args:
+        request(googleapiclient.http.HttpRequest): The Youtube API execute request to process
+        max_retries(int): Maximum # of times to retry an upload (default 10)
+
+    Returns:
+        dict: The YouTube API response
+    """
+    response = None
+    error = None
+
+    retry = 0
+    retry_exceptions = (OSError, http.client.HTTPException)
+    retry_statuses = [500, 502, 503, 504]
+
+    while response is None:
+        try:
+            _, response = request.next_chunk()
+            if response is None or 'id' not in response:
+                raise Exception('YouTube upload failed: %s' % response)
+        except HttpError as e:
+            if e.resp.status in retry_statuses:
+                error = e
+            else:
+                raise
+        except retry_exceptions as e:
+            error = e
+
+        if error is not None:
+            retry += 1
+            if retry > max_retries:
+                log.error('Final upload failure', exc_info=error)
+                raise Exception('Retried YouTube upload 10x, giving up') from error
```

Comment:
Can you create a subclass of `Exception` here to make the exception easier to catch?

Reply:
OK